File: 34.find-first-and-last-position-of-element-in-sorted-array.py
#
# @lc app=leetcode id=34 lang=python3
#
# [34] Find First and Last Position of Element in Sorted Array
#
class Solution:
    def searchRange(self, nums, target):
        if not nums: return [-1, -1]
        
        t_index = -1
        start, end = 0, len(nums) - 1
        while start <= end:
            mid = start + (end - start) // 2
            if nums[mid] == target:
                t_index = mid
                break
            elif nums[mid] < target:
                start = mid + 1
            else:
                end = mid - 1

        if t_index == -1:
            return [-1, -1]
        # print(t_index)
        min_i, max_i = 0, len(nums) - 1

        i = t_index
        while i > 0:
            if nums[i-1] != nums[i]:
                min_i = i
                break
            i -= 1
        
        j = t_index
        while j < len(nums) - 1:
            if nums[j+1] != nums[j]:
                max_i = j
                break
            j += 1


        return [min_i, max_i]



s = Solution()
# print(s.searchRange([5,7,7,8,8,10], 6))
print(s.searchRange([5,7,7,8,8,10], 8))

# 4pm
# binary searh
# find the target index

        

File: 34.find-first-and-last-position-of-element-in-sorted-array.java
/*
 * @lc app=leetcode id=34 lang=java
 *
 * [34] Find First and Last Position of Element in Sorted Array
 */

// @lc code=start
class Solution {
    public int[] searchRange(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return new int[]{-1, -1};
        }
        
        int left = searchFirst(nums, target);
        int right = searchLast(nums, target);
        
        return new int[]{left, right};
    }
    
    private int searchFirst(int[] A, int x) {
        int start = 0;
        int end = A.length - 1;
        
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            
            if (A[mid] >= x) {
                end = mid;
            } else {
                start = mid;
            }
        }
        if (A[start] == x) {
            return start;
        }
        if (A[end] == x) {
            return end;
        }
        return -1;
    }
    
    private int searchLast(int[] A, int x) {
        int start = 0;
        int end = A.length - 1;
        
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            
            if (A[mid] > x) {
                end = mid;
            } else {
                start = mid;
            }
        }
        if (A[end] == x) {
            return end;
        }
        
        if (A[start] == x) {
            return start;
        }
        return -1;
    }
}
// @lc code=end

